E2E: migrate the `Editor: Schedule a Post` spec.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR migrates the Editor: Schedule a Post spec to Playwright.

Key changes:
- addition of new methods and selectors to in the `EditorSettingsSidebarComponent` to interact with the scheduler.
- changes to the `GutenbergEditorPage.publish` method to support both Schedule and regular Post interactions.

#### Testing instructions

- [x] Ensure the following:
  - [x] Desktop passes.
  - [x] Mobile passes.

Related to #55950.

worldomonation committed:
Update POM to support interaction to schedule a post.

worldomonation committed:
Implement Editor: Schedule Post test.

worldomonation committed:
Remove errant pause

worldomonation committed:
Close the EditorSettingsSidebarComponent.

dpasque commented on packages/calypso-e2e/src/lib/components/editor-settings-sidebar-component.ts:
```diff
@@ -134,7 +150,71 @@ export class EditorSettingsSidebarComponent {
-	 * Clikcks on the Revisions section in the sidebar to show a revisions modal.
+	 * Opens the scheduler if required.
+	 */
+	async openSchedule(): Promise< void > {
```
I really like how you designed these methods. They are clear about the end state, and resilient to getting it there regardless of the starting state. This is so much cleaner than a "toggle" based approach!

dpasque commented on packages/calypso-e2e/src/lib/components/editor-settings-sidebar-component.ts:
```diff
@@ -134,7 +150,71 @@ export class EditorSettingsSidebarComponent {
+	 * @param {Schedule} date Date of the article to be scheduled.
+	 */
+	async schedule( date: Schedule ): Promise< void > {
+		for ( const [ key, value ] of Object.entries( date ) ) {
```
100% optional as this works and reads well, but there is actually a kinda nifty way you can get type safety on these keys and values! Which could be potentially useful because logic is based on the key value, and the interface has different types for values:

```typescript
		let key: keyof Schedule;
		for ( key in date ) {
			if ( key === 'meridian' ) {
				// am/pm is a button.
				await this.frame.click( selectors.scheduleMeridianButton( date[ key ] ) );
				continue;
			}
			if ( key === 'month' ) {
				// For month numbers less than 10, pad the digit to be
				// 2 digits as required by the select.
				await this.frame.selectOption(
					'select[name="month"]',
					date[ key ].toString().padStart( 2, '0' )
				);
				continue;
			}

			// Regular input fields.
			const targetSelector = selectors.scheduleInput( key );
			await this.frame.fill( targetSelector, date[ key ].toString() );
		}
```

What's cool about this is it would catch if you misspelled `meridian`, and it also would stop you if you tried to call `scheduleMeridianButton` while in the `key === 'month'` block!

Totally optional, but check it out and see if you like it!

worldomonation replied:
Changed in 2772e4a01442d2997d83025ea17b5806b83a77f6.

dpasque commented on test/e2e/specs/specs-wpcom/wp-editor__schedule.ts:
```diff
@@ -0,0 +1,152 @@
+			const currentDate = new Date();
+			await editorSettingsSidebarComponent.schedule( {
+				year: currentDate.getFullYear() + 1,
+				month: ( ( new Date().getMonth() + 1 ) % 12 ) + 1,
```
Is the modulo and extra addition necessary here? I may not be following this exactly, but shouldn't we be able to just do `currentDate.getMonth() + 1`? This would give parity with how we approach this below in the past date section!

worldomonation replied:
It seems like modulo is required to generate a 0-11 range for the month:

```
> ( ( new Date("2022-01-01").getMonth() + 1 )  ) + 1
13
> ( ( new Date("2022-01-01").getMonth() + 1 ) % 12 ) + 1
1
```

With that said, I re-evaluated this portion in light of your subsequent comment below, and I think I was trying to solve a non-existent problem. I've since removed this part in 2772e4a01442d2997d83025ea17b5806b83a77f6.

dpasque commented on test/e2e/specs/specs-wpcom/wp-editor__schedule.ts:
```diff
@@ -0,0 +1,152 @@
+		} );
+
+		it( 'Schedule the post for next year', async function () {
+			const currentDate = new Date();
+			await editorSettingsSidebarComponent.schedule( {
+				year: currentDate.getFullYear() + 1,
```
Very small nit, but it took me a couple read throughs to make sure I understood all the date logic here. I liked the approach you did in the past section where you took a date and modified it to be what you wanted, and then called the schedule function with the corresponding parts. What do you think of doing that here too for parity?

So, e.g.
```typescript
const futureDate = new Date();
futureDate.setFullYear( new Date().getFullYear() + 1);
```

worldomonation replied:
I've modified this part to first define the date object, then use the date object to set the params.

dpasque reviewed (approved):
A few small nits and a typing suggestion, but otherwise good to go!

You did a really nice job designing this spec and the methods on the POM classes -- it's really clean, deterministic, and easy to understand what is happening! Nice work! 😄

worldomonation committed:
Changes:
- use type-checked iteration through the keys of `date`.
- change how future date was being set.

worldomonation committed:
Change dates used in test
